### PR TITLE
[Feature] Support PERCENT_RANK window function 

### LINKED ISSUE #13021: NTH_VALUE windows function  support
NTH_VALUE function not support

https://dev.mysql.com/doc/refman/8.0/en/window-function-descriptions.html
https://dev.mysql.com/doc/refman/8.0/en/window-function-descriptions.html#function_nth-value
the standard Windows function current not support as fellowing：
CUME_DIST()
NTH_VALUE()
PERCENT_RANK()

and **medium()** function in the windows function

### PR DESCRIPTION
**Pull request by Moonm3n**
## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [x] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #13021 #19632

## Problem Summary(Required):

The PERCENT_RANK() function calculates the relative rank of a row within a result set as a percentage. It returns the percentage of partition values less than the value in the current row, excluding the highest value. The return values range from 0 to 1. This function is useful for percentile calculations and analyzing data distribution.

The PERCENT_RANK() function is calculated using the following formula, where rank represents the row rank and rows represents the number of partition rows:

~~~Plain Text
(rank - 1) / (rows - 1)
~~~

It can not been used with the windowing clause. And for the implementation, it uses `RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`.

NOTE that, PERCENT_RANK need the boundary of partition to calculate its value. For these functions, we must wait util the partition finished. Therefore, it uses `_process_by_partition_for_unbounded_preceding_range_frame_materializing`.

The PERCENT_RANK window function is only supported by the pipeline engine. BE will report the NotSupported error, when non-pipeline engine executes PERCENT_RANK in BE.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
- [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: docs/sql-reference/sql-functions/Window_function.md [hunk 1]
```diff
@@ -691,6 +731,50 @@ from int_t
 where property in ('prime','square');
 ~~~
 
+### PERCENT_RANK()
+
+The PERCENT_RANK() function calculates the relative rank of a row within a result set as a percentage. It returns the percentage of partition values less than the value in the current row, excluding the highest value. The return values range from 0 to 1. This function is useful for percentile calculations and analyzing data distribution.
+
+The PERCENT_RANK() function is calculated using the following formula, where rank represents the row rank and rows represents the number of partition rows:
+
```
Comment (fzhedu) [inline docs/sql-reference/sql-functions/Window_function.md:739]:
```
This function should be used with ORDER BY to sort partition rows into the desired order. Without ORDER BY, all rows are peers.

```

File: docs/sql-reference/sql-functions/Window_function.md [hunk 2]
```diff
@@ -691,6 +731,50 @@ from int_t
...
+            ORDER BY y
+        ) AS `percent_rank`
+FROM int_t;
+~~~
+
+~~~Plain Text
++---+---+-----------------------+
+| x | y | percent_rank          |
++---+---+-----------------------+
+| 1 | 1 | 0.0000000000000000    |
```
Comment (mofeiatwork) [inline docs/sql-reference/sql-functions/Window_function.md:766]: shall we reduce the precision? like `0.12` ? 
what's the behavior of other dbs ?

**Review by murphyatwork — DISMISSED** [review]

File: docs/sql-reference/sql-functions/Window_function.md (hunk 2 above)
Comment (Moonm3n) [reply docs/sql-reference/sql-functions/Window_function.md:766]: MySQL maintains precision up to 16 decimal places. However, when the value is 0.0000000000000000, MySQL displays it as 0. As for this PR, the code logic is accurate, I will fix the document.

Comment (wangsimo0) [reply docs/sql-reference/sql-functions/Window_function.md:766]: hi @Moonm3n , could you please rebase the code?

File: docs/sql-reference/sql-functions/Window_function.md (hunk 1 above)
Comment (Moonm3n) [reply docs/sql-reference/sql-functions/Window_function.md:743]: Done.

File: docs/sql-reference/sql-functions/Window_function.md (hunk 2 above)
Comment (Moonm3n) [reply docs/sql-reference/sql-functions/Window_function.md]: > 

Done, PTAL.

File: be/src/exec/analytor.cpp [hunk 3]
```diff
@@ -140,9 +140,10 @@ Status Analytor::prepare(RuntimeState* state, ObjectPool* pool, RuntimeProfile*
             _need_partition_materializing = true;
         }
 
-        if (fn.name.function_name == "cume_dist") {
+        if (fn.name.function_name == "cume_dist" || fn.name.function_name == "percent_rank") {
             if (!state->enable_pipeline_engine()) {
-                return Status::NotSupported("The CUME_DIST window function is only supported by the pipeline engine.");
+                return Status::NotSupported(strings::Substitute(
+                        "The $0 window function is only supported by the pipeline engine.", fn.name.function_name));
             }
```
Comment (satanson) [inline be/src/exec/analytor.cpp:147]: it is not very clear
so can we introduce a function "should_set_partition_size" to tacke with functions like percent_rank and dist_cume?

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 9 / 9 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 9 / 9 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|2|2|100.00%|[]|

File: be/src/exec/analytor.cpp (hunk 3 above)
Comment (Moonm3n) [reply be/src/exec/analytor.cpp:147]: Done, I use `require_partition_size(fn.name.function_name)` to tackle with functions like percent_rank and dist_cume. `_should_set_partition_size` to indicate whether partition size needs to be set.
PTAL.

**Review by satanson — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**